discovery_engine_core/bindings/src/lib.rs

```diff
@@ -31,8 +31,83 @@ pub mod types;
 
 #[async_bindgen::api]
 impl AsyncCore {
-    /// Adds two bytes.
-    pub async fn add(x: u8, y: u8) -> u8 {
-        x + y
+    /// Initializes the engine.
+    pub async fn initialize(
+        config: Box<core::InitConfig>,
+        state: &Option<Vec<u8>>,
```

Comment:
Iff we always drop the vec in the calling code then having a `Option<Vec<u8>>` would be easier as then rust handles the dropping of it.

Reply:
Also `Option<&Vec<u8>>` means we don't need to have `Option<Vec<u8>>` specific methods, as it's guaranteed nullptr optimized.

Reply:
And with `Option<Box<[u8]>>` or ( `Option<&[u8]>`) we would already have all necessary FFI functions in place.

Reply:
Or  `ptr: *mut u8, len: usize` with nullptr support for `None`.

Reply:
reagrding pointer arguments, i wasn't sure if we can use them at all. as we would have to do unsafe conversions in the function body, making the whole function unsafe, is this supported by `async-bindgen`? and what would be the way you intended `async-bindgen` to be used with, pointer or reference arguments?

Reply:
> Also Option<&Vec<u8>> means we don't need to have Option<Vec<u8>> specific methods, as it's guaranteed nullptr optimized.

i'll go with this for now.

> Iff we always drop the vec in the calling code then having a Box<Option<Vec<u8>>> would be easier as then rust handles the dropping of it.

i don't know how this will be handled on the dart side, might make sense to change to this later on.

Reply:
The idea of async-bindgen is to add `async` to a normal C FFI, so sure `unsafe` is supported and meant to be used where necessary.

The only problem could be `Send`, as pointers are by default not send :thinking: .

So for simplicity lets go with `Option<&Vec<u8>>`.

Reply:
Or `Option<Box<Vec<u8>>>`